agents/dqn.py

```diff
@@ -1,137 +1,132 @@
                 bag_ids=bag_ids,
-                node_idxs=current_node_idx,
+                node_idxs=torch.full([batch_size], self._node_id),
                 extra_infos=zip(
-                    current_node_idx,
+                    torch.unsqueeze(current_node_idx, dim=1),
                     neighbor_node_ids,
-                    destination_node_idx,
-                    next_neighbor_idx
+                    neighbors_q,
+                    torch.unsqueeze(destination_node_idx, dim=1),
+                    torch.unsqueeze(next_neighbor_ids, dim=1)
                 )
             )
-        return {
-            'predicted_next_node_idx': next_neighbor_idx,
-            'predicted_next_node_q': next_neighbor_q,
-        }
+        inputs[self._output_prefix] = next_neighbor_ids
+        inputs.update({
+            'predicted_next_node_idx': next_neighbor_ids,
+            'predicted_next_node_q': neighbors_q,
+        })
+        return inputs
 
-    def learn(self):
-        for trajectory in self._bag_trajectory_memory.sample_trajectories_for_node_idx(
-                self._node_idx,
-                self._trajectory_sample_size
-        ):
+    def learn(self) -> Optional[Tensor]:
+        loss = 0
+        learn_trajectories = self._bag_trajectory_memory.sample_trajectories_for_node_idx(
+            self._node_id,
+            self._trajectory_sample_size,
+            1
+        )
+        if not learn_trajectories:
+            return None
+        for trajectory in learn_trajectories:
             target = trajectory[0]['reward']
-            current_node_idx, neighbor_node_ids, destination_node_idx, next_neighbor = trajectory[0]['extra_infos']
-            if len(trajectory) > 0:
-                current_node_idx_, neighbor_node_ids_, destination_node_idx_, _ = trajectory[1]['extra_infos']
-                neighbor_q_ = self._q_network(
-                    current_node_idx=current_node_idx_,
-                    neighbor_node_ids=neighbor_node_ids_,
-                    destination_node_idx=destination_node_idx_
-                )
-                target += self._discount_factor * neighbor_q_
-            neighbor_q = self._q_network(
+            current_node_idx, neighbor_node_ids, _, destination_node_idx, next_neighbor = trajectory[0]['extra_info']
+            if len(trajectory) > 1:
+                _, neighbor_node_ids_, neighbor_q_, _, next_neighbor_ = trajectory[1]['extra_info']
+                target += self._discount_factor * neighbor_q_[neighbor_node_ids_.padded_values == next_neighbor_].detach()
+            _, neighbor_q = self._q_network(
                 current_node_idx=current_node_idx,
                 neighbor_node_ids=neighbor_node_ids,
                 destination_node_idx=destination_node_idx
             )
-            loss = (target.detach() - neighbor_q[next_neighbor]) ** 2
+            loss += (target - neighbor_q[neighbor_node_ids.padded_values == next_neighbor]) ** 2
```

Comment:
Can you please place here `TODO` for me to check the formulas? Right now I am not quite sure that `.detach()` is used in the correct way. Have to compare it with the paper but since it works right now we can submit it. Just please re-check that the written behavior is exactly working as intended.

Reply:
ok)